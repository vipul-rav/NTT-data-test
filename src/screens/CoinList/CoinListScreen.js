import React, { memo, useEffect, useState } from 'react';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Alert from 'react-bootstrap/Alert';
import Pagination from 'react-bootstrap/Pagination';
import { CoinListTable } from './CoinListTable';
import { LoaderComponent } from '../../components/Loader';
import { API_BASE_URL } from '../../constant/stringConstant';

const CoinListScreen = memo(() => {
  const [coinList, updateCoinList] = useState([]);
  const [activePage, updateActivePage] = useState(1);
  //show loader- loader can handle with middleware and redux globally in App
  const [loading, updateLoading] = useState(true);

  useEffect(() => {
    //call coin List API.
    const getCoinList = async () => {
      try {
        const result = await fetch(
          `${API_BASE_URL}/api/v3/coins/markets?vs_currency=eur&order=market_cap_desc&per_page=10&page=${activePage}&sparkline=false`
        );
        const response = await result.json();
        updateCoinList(response);
        updateLoading(false);
      } catch (e) {
        updateLoading(false);
      }
    };
    getCoinList();
    return () => {};
  }, [activePage]);

  const pageChanged = (page) => {
    const nextPage = activePage + page;
    //no pagination will change if page is less than 0. No Max page size availbele hence keep it as is
    nextPage > 0 && updateActivePage(nextPage);
  };

  if (loading) {
    return <LoaderComponent />;
  }

  return (
    <Container>
      <Row>
        <Col>
          <strong>Coin List</strong>
        </Col>
      </Row>
      <Row>
        <Col>
          {coinList.error ? <Alert variant="danger">{coinList.error}</Alert> : <CoinListTable coinList={coinList} />}
        </Col>
      </Row>
      <Row>
        <Col>
          <Pagination>
            <Pagination.Item onClick={() => pageChanged(-1)}>Prev</Pagination.Item>
            <Pagination.Item onClick={() => pageChanged(1)}>Next</Pagination.Item>
          </Pagination>
        </Col>
      </Row>
    </Container>
  );
});

export { CoinListScreen };
