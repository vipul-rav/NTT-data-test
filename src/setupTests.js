import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

const AllTheProviders = ({ children }) => {
  return <BrowserRouter>{children}</BrowserRouter>;
};
const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
