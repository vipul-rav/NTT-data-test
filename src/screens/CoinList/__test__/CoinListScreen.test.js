import React from 'react';
import { waitForElementToBeRemoved, waitForElement, cleanup, fireEvent } from '@testing-library/react';
import { render } from '../../../setupTests';
import { CoinListScreen } from '../CoinListScreen';

const setup = (overridesProps) => {
  const props = {
    ...overridesProps
  };

  const { container, getByText } = render(<CoinListScreen {...props} />);

  return {
    props,
    container,
    getByText
  };
};

describe('CoinListScreen', () => {
  afterEach(cleanup);
  it('should render', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });
  it('should render', async () => {
    const { container, getByText } = setup();
    const res = await waitForElementToBeRemoved(() => getByText('Please Wait...'));

    expect(container).toMatchSnapshot();
  });
  it('Click Prev button', async () => {
    const { getByText } = setup();
    const res = await waitForElementToBeRemoved(() => getByText('Please Wait...'));
    const btn = getByText('Prev');
    fireEvent.click(btn, -1);
  });
  it('click Prev button for negative value', async () => {
    const { getByText } = setup();
    const res = await waitForElementToBeRemoved(() => getByText('Please Wait...'));
    const btn = getByText('Prev');
    fireEvent.click(btn, -2);
  });
  it('CLick Next button', async () => {
    const { getByText } = setup();
    const res = await waitForElementToBeRemoved(() => getByText('Please Wait...'));
    const btn = getByText('Next');
    fireEvent.click(btn, 1);
  });
});
