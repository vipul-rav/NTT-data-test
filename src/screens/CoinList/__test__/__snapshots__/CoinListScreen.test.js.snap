// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinListScreen should render 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="spinner-border"
        />
        Please Wait...
      </div>
    </div>
  </div>
</div>
`;

exports[`CoinListScreen should render 2`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <strong>
          Coin List
        </strong>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <table
          class="table table-striped table-bordered"
        >
          <thead>
            <tr>
              <th>
                Image
              </th>
              <th>
                Name
              </th>
              <th>
                Symbol
              </th>
              <th>
                Current Price
              </th>
              <th>
                High Price
              </th>
              <th>
                Low Price
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                <img
                  alt="Bitcoin"
                  height="24"
                  src="https://assets.coingecko.com/coins/images/1/large/bitcoin.png?1547033579"
                  width="24"
                />
              </td>
              <td>
                <a
                  href="/coin-details/bitcoin"
                >
                  Bitcoin
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/bitcoin"
                >
                  btc
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/bitcoin"
                >
                  20408
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/bitcoin"
                >
                  20602
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/bitcoin"
                >
                  20217
                </a>
              </td>
            </tr>
            <tr>
              <td>
                <img
                  alt="Ethereum"
                  height="24"
                  src="https://assets.coingecko.com/coins/images/279/large/ethereum.png?1595348880"
                  width="24"
                />
              </td>
              <td>
                <a
                  href="/coin-details/ethereum"
                >
                  Ethereum
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/ethereum"
                >
                  eth
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/ethereum"
                >
                  1389.04
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/ethereum"
                >
                  1399.7
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/ethereum"
                >
                  1353.23
                </a>
              </td>
            </tr>
            <tr>
              <td>
                <img
                  alt="Tether"
                  height="24"
                  src="https://assets.coingecko.com/coins/images/325/large/Tether-logo.png?1598003707"
                  width="24"
                />
              </td>
              <td>
                <a
                  href="/coin-details/tether"
                >
                  Tether
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/tether"
                >
                  usdt
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/tether"
                >
                  1.02
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/tether"
                >
                  1.023
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/tether"
                >
                  1.007
                </a>
              </td>
            </tr>
            <tr>
              <td>
                <img
                  alt="BNB"
                  height="24"
                  src="https://assets.coingecko.com/coins/images/825/large/bnb-icon2_2x.png?1644979850"
                  width="24"
                />
              </td>
              <td>
                <a
                  href="/coin-details/binancecoin"
                >
                  BNB
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/binancecoin"
                >
                  bnb
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/binancecoin"
                >
                  299.09
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/binancecoin"
                >
                  300.34
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/binancecoin"
                >
                  295.86
                </a>
              </td>
            </tr>
            <tr>
              <td>
                <img
                  alt="USD Coin"
                  height="24"
                  src="https://assets.coingecko.com/coins/images/6319/large/USD_Coin_icon.png?1547042389"
                  width="24"
                />
              </td>
              <td>
                <a
                  href="/coin-details/usd-coin"
                >
                  USD Coin
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/usd-coin"
                >
                  usdc
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/usd-coin"
                >
                  1.02
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/usd-coin"
                >
                  1.021
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/usd-coin"
                >
                  1.006
                </a>
              </td>
            </tr>
            <tr>
              <td>
                <img
                  alt="XRP"
                  height="24"
                  src="https://assets.coingecko.com/coins/images/44/large/xrp-symbol-white-128.png?1605778731"
                  width="24"
                />
              </td>
              <td>
                <a
                  href="/coin-details/ripple"
                >
                  XRP
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/ripple"
                >
                  xrp
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/ripple"
                >
                  0.504432
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/ripple"
                >
                  0.508417
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/ripple"
                >
                  0.494313
                </a>
              </td>
            </tr>
            <tr>
              <td>
                <img
                  alt="Binance USD"
                  height="24"
                  src="https://assets.coingecko.com/coins/images/9576/large/BUSD.png?1568947766"
                  width="24"
                />
              </td>
              <td>
                <a
                  href="/coin-details/binance-usd"
                >
                  Binance USD
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/binance-usd"
                >
                  busd
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/binance-usd"
                >
                  1.021
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/binance-usd"
                >
                  1.023
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/binance-usd"
                >
                  1.004
                </a>
              </td>
            </tr>
            <tr>
              <td>
                <img
                  alt="Cardano"
                  height="24"
                  src="https://assets.coingecko.com/coins/images/975/large/cardano.png?1547034860"
                  width="24"
                />
              </td>
              <td>
                <a
                  href="/coin-details/cardano"
                >
                  Cardano
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/cardano"
                >
                  ada
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/cardano"
                >
                  0.438142
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/cardano"
                >
                  0.441386
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/cardano"
                >
                  0.433476
                </a>
              </td>
            </tr>
            <tr>
              <td>
                <img
                  alt="Solana"
                  height="24"
                  src="https://assets.coingecko.com/coins/images/4128/large/solana.png?1640133422"
                  width="24"
                />
              </td>
              <td>
                <a
                  href="/coin-details/solana"
                >
                  Solana
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/solana"
                >
                  sol
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/solana"
                >
                  34.57
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/solana"
                >
                  34.76
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/solana"
                >
                  34.07
                </a>
              </td>
            </tr>
            <tr>
              <td>
                <img
                  alt="Dogecoin"
                  height="24"
                  src="https://assets.coingecko.com/coins/images/5/large/dogecoin.png?1547792256"
                  width="24"
                />
              </td>
              <td>
                <a
                  href="/coin-details/dogecoin"
                >
                  Dogecoin
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/dogecoin"
                >
                  doge
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/dogecoin"
                >
                  0.065352
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/dogecoin"
                >
                  0.067188
                </a>
              </td>
              <td>
                <a
                  href="/coin-details/dogecoin"
                >
                  0.064936
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <ul
          class="pagination"
        >
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
              role="button"
              tabindex="0"
            >
              Prev
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
              role="button"
              tabindex="0"
            >
              Next
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
